sensor/kubernetes/listener/resources/networkpolicy.go

```diff
@@ -2,41 +2,99 @@ package resources
+	"github.com/stackrox/rox/pkg/features"
 	networkPolicyConversion "github.com/stackrox/rox/pkg/protoconv/networkpolicy"
+	"github.com/stackrox/rox/sensor/common/detector"
 	"github.com/stackrox/rox/sensor/common/store"
 	networkingV1 "k8s.io/api/networking/v1"
 )
 
 // networkPolicyDispatcher handles network policy resource events.
 type networkPolicyDispatcher struct {
-	store store.NetworkPolicyStore
+	netpolStore     store.NetworkPolicyStore
+	deploymentStore *DeploymentStore
+	detector        detector.Detector
 }
 
-func newNetworkPolicyDispatcher(nps store.NetworkPolicyStore) *networkPolicyDispatcher {
+func newNetworkPolicyDispatcher(networkPolicyStore store.NetworkPolicyStore, deploymentStore *DeploymentStore, detector detector.Detector) *networkPolicyDispatcher {
 	return &networkPolicyDispatcher{
-		store: nps,
+		netpolStore:     networkPolicyStore,
+		deploymentStore: deploymentStore,
+		detector:        detector,
 	}
 }
 
-// ProcessEvent processes a network policy resource event, and returns the sensor events to generate.
-func (h *networkPolicyDispatcher) ProcessEvent(newObj, _ interface{}, action central.ResourceAction) []*central.SensorEvent {
-	np := newObj.(*networkingV1.NetworkPolicy)
-	netPolicy := networkPolicyConversion.KubernetesNetworkPolicyWrap{NetworkPolicy: np}.ToRoxNetworkPolicy()
+// Process processes a network policy resource event, and returns the sensor events to generate.
+func (h *networkPolicyDispatcher) ProcessEvent(obj, old interface{}, action central.ResourceAction) []*central.SensorEvent {
+	np := obj.(*networkingV1.NetworkPolicy)
 
-	switch action {
-	case central.ResourceAction_CREATE_RESOURCE, central.ResourceAction_UPDATE_RESOURCE:
-		h.store.Upsert(netPolicy)
-	case central.ResourceAction_REMOVE_RESOURCE:
-		h.store.Delete(netPolicy.GetId(), netPolicy.GetNamespace())
+	roxNetpol := networkPolicyConversion.KubernetesNetworkPolicyWrap{NetworkPolicy: np}.ToRoxNetworkPolicy()
+
+	if features.NetworkPolicySystemPolicy.Enabled() {
+		var roxOldNetpol *storage.NetworkPolicy
+		if old != nil {
+			oldNp := old.(*networkingV1.NetworkPolicy)
+			roxOldNetpol = networkPolicyConversion.KubernetesNetworkPolicyWrap{NetworkPolicy: oldNp}.ToRoxNetworkPolicy()
+		}
+		sel, matchesAll := h.getSelector(roxNetpol, roxOldNetpol, action)
+		if action == central.ResourceAction_REMOVE_RESOURCE {
+			h.netpolStore.Delete(roxNetpol.GetId(), roxNetpol.GetNamespace())
+		} else {
+			h.netpolStore.Upsert(roxNetpol)
+		}
+
+		h.updateDeploymentsFromStore(roxNetpol, sel, matchesAll)
 	}
 
 	return []*central.SensorEvent{
 		{
 			Id:     string(np.UID),
 			Action: action,
 			Resource: &central.SensorEvent_NetworkPolicy{
-				NetworkPolicy: networkPolicyConversion.KubernetesNetworkPolicyWrap{NetworkPolicy: np}.ToRoxNetworkPolicy(),
+				NetworkPolicy: roxNetpol,
 			},
 		},
 	}
 }
+
+func (h *networkPolicyDispatcher) getSelector(np, oldNp *storage.NetworkPolicy, action central.ResourceAction) (selector, bool) {
```

Comment:
I think there is some potential for improvement in this function with regards to returning both a selector and a `matchAll` boolean value. A few things that I thought:

1. I understand that `nothingSelector` will not match anything if used as a matcher, but couldn't we simply check introduce a new method (as opposed to `SelectorFromMap`) that will return `labels.Everything()` instead?
2. If old policy is a match-all policy we can early return, since whatever the new policy matches is irrelevant, we'll anyway select all deployments in the namespace.
3. Consider that if no old policy is provided we can start the selector from a `labels.Nothing()` and combine that with `or(...)`. This would simplify the second portion of the code. 

I tried exercising this below but didn't test it since the unit tests are expecting the `matchAll` returns. This comes back to my previous comment about testing internal functions. I understand why you did it like this, but the flip side is that it's difficult to refactor code without having to refactor the tests. 😞 

```go
func matcherOrEverything(labelsMap map[string]string) labels.Selector {
	if len(labelsMap) == 0 {
		return labels.Everything()
	}
	return labels.SelectorFromSet(labels.Set(labelsMap))
}

func (h *networkPolicyDispatcher) getSelector(np, oldNp *storage.NetworkPolicy, action central.ResourceAction) selector {
	var sel selector
	if oldNp != nil {
		sel = matcherOrEverything(oldNp.GetSpec().GetPodSelector().GetMatchLabels())	
                // potential early return but not necessary
		if sel == labels.Everything() {
			return sel
		}	
	} else {
		sel = labels.Nothing()
	}

	matchLabels := np.GetSpec().GetPodSelector().GetMatchLabels()
	if action == central.ResourceAction_UPDATE_RESOURCE {
		sel = or(sel, matcherOrEverything(matchLabels))
	} else if action == central.ResourceAction_CREATE_RESOURCE {
		sel = matcherOrEverything(matchLabels)
	}	
	
	return sel
}
```

Reply:
It could even be simpler if we ignore the `action` check, which won't make a difference if we're combining with `or labels.Nothing()`:

```go
func (h *networkPolicyDispatcher) getSelector(np, oldNp *storage.NetworkPolicy, action central.ResourceAction) selector {
	var sel selector
	if oldNp != nil {
		sel = matcherOrEverything(oldNp.GetSpec().GetPodSelector().GetMatchLabels())	
                // potential early return but not necessary
		if sel == labels.Everything() {
			return sel
		}	
	} else {
		sel = labels.Nothing()
	}

	matchLabels := np.GetSpec().GetPodSelector().GetMatchLabels()
        return or(sel, matcherOrEverything(matchLabels))
}
```